syntax = "proto3";

package auth;

// Authentication service definition
service AuthService {
  // Login with username and password
  rpc Login(LoginRequest) returns (LoginResponse);
  
  // Validate a session token
  rpc ValidateSession(ValidateSessionRequest) returns (ValidateSessionResponse);
  
  // Logout (invalidate session)
  rpc Logout(LogoutRequest) returns (LogoutResponse);
  
  // Health check - returns if server is alive and config timestamp
  rpc IsAlive(IsAliveRequest) returns (IsAliveResponse);
  
  // Get process PID
  rpc GetPID(GetPIDRequest) returns (GetPIDResponse);
  
  // List all users (for management UI)
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
  
  // Add a new user
  rpc AddUser(AddUserRequest) returns (AddUserResponse);
  
  // Delete a user
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
  
  // Issue a new JWT token for a user
  rpc IssueToken(IssueTokenRequest) returns (IssueTokenResponse);
  
  // Get JWT token information by JTI
  rpc GetTokenInfo(GetTokenInfoRequest) returns (GetTokenInfoResponse);
  
  // Database management
  rpc GetDatabaseList(GetDatabaseListRequest) returns (GetDatabaseListResponse);
  rpc AddDatabase(AddDatabaseRequest) returns (AddDatabaseResponse);
  rpc RemoveDatabase(RemoveDatabaseRequest) returns (RemoveDatabaseResponse);
  rpc UpdateDatabase(UpdateDatabaseRequest) returns (UpdateDatabaseResponse);
  rpc ChangeCurrentDatabase(ChangeCurrentDatabaseRequest) returns (ChangeCurrentDatabaseResponse);
}

// Login request message
message LoginRequest {
  string username = 1;
  string password = 2;
}

// Login response message
message LoginResponse {
  bool success = 1;
  string message = 2;
  string session_token = 3;  // JWT or session ID
  int64 expires_at = 4;       // Unix timestamp
}

// Validate session request
message ValidateSessionRequest {
  string session_token = 1;
}

// Validate session response
message ValidateSessionResponse {
  bool valid = 1;
  string username = 2;
  string message = 3;
}

// Logout request
message LogoutRequest {
  string session_token = 1;
}

// Logout response
message LogoutResponse {
  bool success = 1;
  string message = 2;
}

// IsAlive request (empty)
message IsAliveRequest {
}

// IsAlive response
message IsAliveResponse {
  bool alive = 1;
  int64 unix_stamp_ms = 2;
}

// GetPID request (empty)
message GetPIDRequest {
}

// GetPID response
message GetPIDResponse {
  int32 pid = 1;
}

// ListUsers request (empty)
message ListUsersRequest {
}

// User info for listing
message UserInfo {
  int32 uid = 1;
  string username = 2;
  string password_hash = 3;
  repeated string jwt_token_ids = 4;  // List of JTI (JWT Token IDs)
}

// ListUsers response
message ListUsersResponse {
  repeated UserInfo users = 1;
}

// AddUser request
message AddUserRequest {
  string username = 1;
  string password = 2;
}

// AddUser response
message AddUserResponse {
  bool success = 1;
  string message = 2;
  int32 uid = 3;  // UID of created user (if successful)
}

// DeleteUser request
message DeleteUserRequest {
  int32 uid = 1;
}

// DeleteUser response
message DeleteUserResponse {
  bool success = 1;
  string message = 2;
}

// IssueToken request
message IssueTokenRequest {
  int32 uid = 1;
}

// IssueToken response
message IssueTokenResponse {
  bool success = 1;
  string message = 2;
  string jti = 3;        // JWT Token ID
  string token = 4;      // The actual JWT token
}

// GetTokenInfo request
message GetTokenInfoRequest {
  string jti = 1;
}

// GetTokenInfo response
message GetTokenInfoResponse {
  bool success = 1;
  string message = 2;
  string jti = 3;
  int32 uid = 4;
  string token = 5;
  int64 created_at = 6;
  int64 expires_at = 7;
}

// Database management messages
message DatabaseInfo {
  int32 id = 1;
  string name = 2;
  string type = 3;
  string host = 4;
  int32 port = 5;
  string database = 6;
  string username = 7;
  string password = 8;
  string path = 9;
  bool is_default = 10;
  bool is_removable = 11;
}

message GetDatabaseListRequest {}

message GetDatabaseListResponse {
  bool success = 1;
  string message = 2;
  repeated DatabaseInfo databases = 3;
  int32 current_database_id = 4;
}

message AddDatabaseRequest {
  string name = 1;
  string type = 2;
  string host = 3;
  int32 port = 4;
  string database = 5;
  string username = 6;
  string password = 7;
  string path = 8;
}

message AddDatabaseResponse {
  bool success = 1;
  string message = 2;
  DatabaseInfo database = 3;
}

message RemoveDatabaseRequest {
  int32 db_id = 1;
}

message RemoveDatabaseResponse {
  bool success = 1;
  string message = 2;
}

message UpdateDatabaseRequest {
  int32 db_id = 1;
  string name = 2;
  string host = 3;
  int32 port = 4;
  string database = 5;
  string username = 6;
  string password = 7;
  string path = 8;
}

message UpdateDatabaseResponse {
  bool success = 1;
  string message = 2;
  DatabaseInfo database = 3;
}

message ChangeCurrentDatabaseRequest {
  int32 db_id = 1;
}

message ChangeCurrentDatabaseResponse {
  bool success = 1;
  string message = 2;
}

